### DISCUSSION TITLE
Overwriting dependency in other file, for testing

### DISCUSSION THREAD

**Discussion by nicolai-b** [fastapi/fastapi#6708] [Questions] [answered]
Hello,
I have added authentication to use my API functions, so it requires the user to log in to be able to use them. However, this is a problem when trying to test my functions, as the test client is not authenticated and returns code 401. I am trying to override my dependency so the test client will not have to be authenticated, but it does not work when I run ```pytest```. A lot of my code is similar to the FastAPI security example. Does anyone know why my dependency isn't overwritten?

Running ```pytest``` returns ```FAILED test_main.py::test_read_status - assert 401 == 200```

### main.py:
```Python
from fastapi import FastAPI
from fastapi.security import OAuth2PasswordBearer

app = FastAPI()
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

async def get_current_user(token: str = Depends(oauth2_scheme)):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(users_db, username=token_data.username)
    if user is None:
        raise credentials_exception
    return user

async def get_current_active_user(current_user: User = Depends(get_current_user)):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user

@app.get("/status")
def get_status(current_user: User = Depends(get_current_active_user)):
    status = get_status()
    return status
```
### test_main.py:
```Python
from fastapi.testclient import TestClient

from main import app
from main import get_current_active_user

client = TestClient(app)

async def override_dependency(q: Optional[str] = None):
    return {"q": q, "skip": 5, "limit": 10}

app.dependency_overrides[get_current_active_user] = override_dependency

def test_read_status():
    response = client.get("/status")
    assert response.status_code == 200

app.dependency_overrides = {}
```

**Comment by STeveShary** [comment, marked as answer]
This is whoops one.   You set your overrides here:
```
app.dependency_overrides[get_current_active_user] = override_dependency
```

but then you remove them before the test is executed here:
```
app.dependency_overrides = {}
```
Just remove the `app.dependency_overrides = {}` line and it should all work.


If you want to setup the app, I would use a test fixture setup with pytest: https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended

**Reply by YuriiMotov** [reply]
Top-level code (which is not inside functions) is executed before tests start. So, you override dependency, then reset it and then run tests.

If you want to override before and remove after test, use fixtures with yield.
Something like this (simplified version)

```python
@pytest.fixture(autouse=True)
def override_deps():
    app.dependency_overrides[get_current_active_user] = override_dependency
    yield
    app.dependency_overrides = {}    
```

**Comment by nicolai-b** [comment]
Ok, thanks! It worked

**Comment by STeveShary** [comment]
@nicolai-b Can you close this issue if it is resolved?
